plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/QueryBuilder.java

```diff
@@ -109,16 +87,48 @@ public PreparedStatement buildSql(
 
         sql += getGroupBy(groupingSets);
 
-        String query = sqlFunction.apply(sql);
-        log.debug("Preparing query: %s", query);
-        PreparedStatement statement = client.getPreparedStatement(connection, query);
+        return new PreparedQuery(sql, accumulator.build());
+    }
 
-        for (int i = 0; i < accumulator.size(); i++) {
-            BoundValue boundValue = accumulator.get(i);
+    public PreparedStatement buildSql(
+            ConnectorSession session,
+            Connection connection,
+            RemoteTableName remoteTableName,
+            Optional<List<List<JdbcColumnHandle>>> groupingSets,
+            List<JdbcColumnHandle> columns,
+            TupleDomain<ColumnHandle> tupleDomain,
+            Optional<String> additionalPredicate,
+            Function<String, String> sqlFunction)
+            throws SQLException
+    {
+        PreparedQuery preparedQuery = prepareQuery(
+                session,
+                connection,
+                remoteTableName,
+                groupingSets,
+                columns,
+                tupleDomain,
+                additionalPredicate);
+        preparedQuery = preparedQuery.transformQuery(sqlFunction);
+        return prepareStatement(session, connection, preparedQuery);
+    }
+
+    public PreparedStatement prepareStatement(
+            ConnectorSession session,
+            Connection connection,
+            PreparedQuery preparedQuery)
+            throws SQLException
+    {
+        log.debug("Preparing query: %s", preparedQuery.getQuery());
+        PreparedStatement statement = client.getPreparedStatement(connection, preparedQuery.getQuery());
+
+        List<QueryParameter> parameters = preparedQuery.getParameters();
+        for (int i = 0; i < parameters.size(); i++) {
+            QueryParameter parameter = parameters.get(i);
             int parameterIndex = i + 1;
-            WriteFunction writeFunction = boundValue.getWriteFunction();
+            WriteFunction writeFunction = getWriteFunction(session, connection, parameter.getJdbcType(), parameter.getType());
             Class<?> javaType = writeFunction.getJavaType();
-            Object value = boundValue.getValue();
+            Object value = parameter.getValue().orElseThrow(() -> new VerifyException("Value is missing"));
```

Comment:
Why `orElse`. Doesn't empty optional mean null? Where is that handled?

Reply:
Huh, yes, empty could mean null, but null never lands here. 
The values are populated from ValueSet (non-nulls) and nullity is handled separately.

Reply:
adding code comment